core/crypto/src/signature.rs

```diff
@@ -556,6 +558,83 @@ impl<'de> serde::Deserialize<'de> for SecretKey {
 #[derive(Clone)]
 pub struct Secp256K1Signature([u8; 65]);
 
+impl Secp256K1Signature {
+    fn check_v(&self) -> Result<u8, crate::errors::ParseSignatureError> {
+        match self.0[64] {
+            27 | 28 => Ok(self.0[64] - 27),
+            _ => Err(crate::errors::ParseSignatureError::InvalidData {
+                error_message: "invalid `v` value".to_string(),
+            }),
+        }
+    }
+
+    pub fn check_signature_values(&self) -> bool {
+        let mut r_bytes = [0u8; 32];
+        r_bytes.copy_from_slice(&self.0[0..32]);
+        let r = U256::from(r_bytes);
+
+        let mut s_bytes = [0u8; 32];
+        s_bytes.copy_from_slice(&self.0[32..64]);
+        let s = U256::from(s_bytes);
+
+        let v = match self.check_v() {
+            Ok(v) => v,
+            Err(_e) => return false,
+        };
+
+        let big_one = U256::one();
+        if r < big_one || s < big_one {
+            return false;
+        }
+
+        let secp256k1_n = U256::from([
+            0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff,
+            0xff, 0xfe, 0xba, 0xae, 0xdc, 0xe6, 0xaf, 0x48, 0xa0, 0x3b, 0xbf, 0xd2, 0x5e, 0x8c,
+            0xd0, 0x36, 0x41, 0x41,
+        ]);
+
+        let secp256k1_half_n = secp256k1_n / U256::from(2);
+        // Reject upper range of s values (ECDSA malleability)
+        if s > secp256k1_half_n {
+            return false;
+        }
+
+        r < secp256k1_n && s < secp256k1_n && (v == 0 || v == 1)
+    }
+
+    pub fn recover(
+        &self,
+        msg: impl AsRef<[u8]>,
```

Comment:
Nit:

```suggestion
        msg: &[u8],
```

Generic functions blow up compile time and binary size. For internal APIs, `AsRef` polymorphism is a net-negative. For public APIs, it sometimes makes sense to wire an `pub fn f(param: impl AsRef<T>)` which is a one-liner delegating to `fn f_(param: &T)`.

Reply:
Actually, why do we need `&[u8]` at all here? Seems like this can take `msg: [u8; 32]`, which will allow changing `Message::from_slice` to `Message::from`, removing one error condition altogether?

Reply:
Ok, then I'll just revert it to that.

Reply:
Fixed: https://github.com/near/nearcore/pull/4380/commits/9c281503ececd4295b166f6b4cbccab1a9b5ab28

I had that before, but then was asked why I didn't add a reference.